### PR TITLE
Create helpers for implementing unary ops whose CUDA implementation is ATen.

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25423 Move the CUDA implementation of trunc to ATen.
* #25372 Move the CUDA implementation of floor to ATen.
* #25285 Move the CUDA implementation of rsqrt to ATen.
* #25041 Move the CUDA implementation of round to ATen.
* **#24879 Create helpers for implementing unary ops whose CUDA implementation is ATen.**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24879

Differential Revision: [D17073557](https://our.internmc.facebook.com/intern/diff/D17073557)

### CODE REVIEW

**Comment by gchanan** [comment]
I'll let @VitalyFedyunin make the final call on this, but I'll express my dislike for these kind of macros:

1) it makes it difficult to figure out if e.g. we ever actually define ceil_ in our code; this has caused a real problem before here: https://github.com/pytorch/pytorch/pull/19938 -- the tradeoff of a few extra lines of boilerplate is not worth it
2) We should consider if there's a common pattern we can apply to all ops -- if we are going to have magic, having one piece of magic that is consistently applied is way better than a ton of magic scattered throughout the code.

**Comment by xuhdev** [comment]
> I'll let @VitalyFedyunin make the final call on this, but I'll express my dislike for these kind of macros:
> 
>     1. it makes it difficult to figure out if e.g. we ever actually define ceil_ in our code; this has caused a real problem before here: #19938 -- the tradeoff of a few extra lines of boilerplate is not worth it
> 

I updated the PR to simplify the helper function and removed the macro. This should have resolved your concern and also helps maintain future uniform changes to all such functions (like this one 6100205eb8a88090581a905e595e7d1d2a8c232d)

>     2. We should consider if there's a common pattern we can apply to all ops -- if we are going to have magic, having one piece of magic that is consistently applied is way better than a ton of magic scattered throughout the code.

I think the helpers I added should work on most unary ops with both CPU and CUDA ATen implementation (which I think will be a fairly large subclass of operators?), as long as they are not as irregular as the one such as `logical_not`.

**Comment by VitalyFedyunin** [comment]
I like current approach (without macros) as it is good trajectory to unify regular, inplace and out operators. Please make sure all tests are passing, also I think it is reasonable to put this on top of UnaryOps.cpp so people can follow example if necessary.

File: aten/src/ATen/native/UnaryOps.cpp [hunk 1]
```diff
@@ -28,32 +28,51 @@
...
 // NOTE:
-// YOU ARE NOT OBLIGED TO USE THESE MACROS
+// YOU ARE NOT OBLIGED TO USE THESE HELPERS
 // If you're writing something more specialized, please don't try to make them
 // work for your case, but just write something new instead.
 
-namespace at {
-namespace native {
+// Helper functions that reduce redundant code in implementing the most typical kind of unary operators. This allows
+// some preprocessing that are unique to some operators (more is forsee-able in the future) and is more flexible and
```
Comment (gchanan) [inline aten/src/ATen/native/UnaryOps.cpp]: nit: foreseeable is a word (doesn't need hyphen) and it implies it's in the future already.

File: aten/src/ATen/native/UnaryOps.cpp [hunk 2]
```diff
@@ -28,32 +28,51 @@
...
+
 // NOTE:
-// YOU ARE NOT OBLIGED TO USE THESE MACROS
+// YOU ARE NOT OBLIGED TO USE THESE HELPERS
 // If you're writing something more specialized, please don't try to make them
 // work for your case, but just write something new instead.
 
-namespace at {
-namespace native {
+// Helper functions that reduce redundant code in implementing the most typical kind of unary operators. This allows
```
Comment (gchanan) [inline aten/src/ATen/native/UnaryOps.cpp]: nit: I'd reorder this a bit.  Basically you want to order of points to be:
1) these are for typical operators
2) therefore, don't add more specialized handling here, they are just for the common case
3) make the preprocessing point.

File: aten/src/ATen/native/UnaryOps.cpp [hunk 3]
```diff
@@ -28,32 +28,51 @@
...
+// some preprocessing that are unique to some operators (more is forsee-able in the future) and is more flexible and
+// elegant than defining a flat fat macro that implements everything.
+template <typename Stub>
+static inline Tensor& unary_op_out_impl(Tensor& result, const Tensor& self, Stub& stub) {
```
Comment (gchanan) [inline aten/src/ATen/native/UnaryOps.cpp]: nit: i'd rename this unary_op_impl_out, keeping "_" and "out" at the end makes it easier to read.

File: aten/src/ATen/native/UnaryOps.cpp [hunk 4]
```diff
@@ -28,32 +28,51 @@
...
-Tensor& bitwise_not_out(Tensor& result, const Tensor& self) {
-  checkBackend("bitwise_not", result, self.type().backend());
-  auto iter = TensorIterator::unary_op(result, self,
-    /*check_mem_overlap=*/true);
-  bitwise_not_stub(iter.device_type(), iter);
-#ifdef BUILD_NAMEDTENSOR
-  at::namedinference::propagate_names(result, self);
-#endif
-  return result;
+Tensor bitwise_not(const Tensor& self) { return unary_op_impl(self, bitwise_not_out); }
```
Comment (gchanan) [inline aten/src/ATen/native/UnaryOps.cpp]: what is bitwise_not_out?  Is it at::bitwise_not_out?  at::native::bitwise_not_out?

**Comment by gchanan** [comment]
ya I agree, the non-macro approach looks good.

**Review by gchanan — APPROVED** [review]
looks good -- sorry for the anti-macro rant :).

BTW did you answer about what the template parameter for e.g. bitwise_not_out becomes?  I assume it's the native::bitwise_not_out, not at::bitwise_not_out, which is different than previous implementations.  Is that the correct thing to use?

**Comment by xuhdev** [comment]
> looks good -- sorry for the anti-macro rant :).
> 

No problem, we all try to work out the best code, both in terms of maintainability and efficiency. I am also a meticulous person myself and I certainly understand this :)

> BTW did you answer about what the template parameter for e.g. bitwise_not_out becomes? I assume it's the native::bitwise_not_out, not at::bitwise_not_out, which is different than previous implementations. Is that the correct thing to use?

It's `native:bitwise_not_out`. I think it is the correct thing. I swapped the order of the definition so it shouldn't be ambiguous to read now.

**Comment by xuhdev** [comment]
@VitalyFedyunin Are you trying to land this PR now? If not, I'll do a quick rebase

**Comment by VitalyFedyunin** [comment]
Go ahead with rebase, but please ping me after it is done.

**Comment by xuhdev** [comment]
@VitalyFedyunin Done!

**Pull request closed without merging**